README.md

```diff
@@ -948,111 +971,140 @@ class C {
+let store = new Store();
+
+container.register('store', store);
+
+let c = container.create(C);
 
-c.foo = 123;
-validate(c);
-// false
+c.store === store; // true
 ```
 
 However, it is not possible to do this as directly _private_ elements, as the key the metadata is defined with cannot be used to access it externally.
 
-This is the purpose of the `access` object that is passed to private elements. This object gives decorators a way to expose access via metadata, like so:
+This is the purpose of the `access` object that is passed to private elements. This object gives decorators a way to expose access via metadata, which can be used in a number of ways. For instance, we can update our dependency injection example to allow injection on private fields:
 
 ```js
-function validatePrivateString(value, context) {
-  let { get } = context.access;
+const INJECTIONS = Symbol();
 
-  context.defineMetadata("validation", (instance) => {
-    let value = get.call(instance);
+function inject(injectionKey) {
+  return (value, context) => {
+    let set = context.access?.set;
 
-    return typeof value === "string";
-  });
+    context.defineMetadata(INJECTIONS, { injectionKey, set })
+  }
 }
 
-function validate(instance) {
-  let metadata = Object.getPrototypeOf(instance)[Symbol.metadata];
+class Container {
+  registry = new Map();
 
-  for (let key in metadata) {
-    let validation = metadata[key].validation;
+  register(injectionKey, value) {
+    this.registry.set(injectionKey, value);
+  }
 
-    if (validation && !validation(instance)) {
-      return false;
-    }
+  lookup(injectionKey) {
+    this.register.get(injectionKey);
   }
 
-  return true;
+  create(Class) {
+    let instance = new Class();
+
+    const { public, private } = instance[Symbol.metadata][INJECTIONS];
+
+    for (let key in public) {
+      instance[key] = this.lookup(public[key].injectionKey);
+    }
+
+    for (let { injectionKey, set } of private) {
+      set.call(instance, this.lookup(injectionKey))
+    }
+
+    return instance;
+  }
 }
 
+class Store {}
+
 class C {
-  @validatePrivateString
-  #foo = "hello!";
+  @inject('store') #store;
 
-  updateFoo(val) {
-    this.#foo = val;
+  get store() {
+    return this.#store;
   }
 }
 
-let c = new C();
-validate(c);
-// true
+let container = new Container();
+let store = new Store();
+
+container.register('store', store);
 
-c.updateFoo(123);
-validate(c);
-// false
+let c = container.create(C);
+
+c.store === store; // true
 ```
 
 Calling the `get` and `set` functions is equivalent to accessing the value on the instance.
 
 ```js
+const FIELD_ACCESS = Symbol();
+
 function exposeField(value, context) {
-  context.defineMetadata("fieldAccess", context.access);
+  context.defineMetadata(FIELD_ACCESS, context.access);
 }
 
 class C {
   @exposeField #x = 1;
 
   updateX() {
-    let { get, set } = C.prototype[Symbol.metadata]["#x"];
+    let { get, set } = this[Symbol.metadata][FIELD_ACCESS].private[0];
```

Comment:
This seems particularly fragile.  If a new private is added or more exist and get rearranged, #x isn't 0 anymore.  Or if C extended another class, and that class had private members, #x isn't 0 anymore.

But again, this is where now having the private name in the decorator can save us.  All we'd need to do is get the privates array from `this`, then get the privates array from `Object.getPrototypeOf(Object.getPrototypeOf(this))`, get the difference to find out which privates are local to this class, then walk through those until we find the metadata value that has been provided with the private name `#x` through its decorator.  Easy.

Reply:
One update I've made here is that `private` metadata is not inherited by default, specifically the make it easier to distinguish between which metadata belongs to which class. This makes updating inheritance easier for users, since that is currently a manual process. I'm not sure if it addresses your concern here though.

I'm not sure how the ordering of the `private` array matters much. For logging purposes, you can include the name from the decorator itself, but that name is meaningless. Really, the important thing is that you get the `access` object which gives you access to private slot for that field/method. Since you cannot override a private value, there shouldn't be any way that these private arrays could collide, or that the order could become important.

Reply:
After more discussion, we decided that the `private` metadata array will inherit via a getter, allowing users to iterate over all of the metadata more easily. This would indeed make it more difficult to get the "own" metadata for a class, but in principle it would still be possible as you've noted. We also don't have any use cases where this would be necessary at the moment, but I'm definitely open to more discussion here if we think there are more use cases in general.